### PR TITLE
KAFKA-8455: Add VoidSerde to Serdes

### PR DESCRIPTION
**Pull request by nizhikov**
As discussed in [KIP-527](https://cwiki.apache.org/confluence/display/KAFKA/KIP-527%3A+Add+NothingSerde+to+Serdes) this PR introduces new VoidSerde.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @ableegoldman @vvcephei

File: clients/src/main/java/org/apache/kafka/common/serialization/VoidDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.serialization;
+
+public class VoidDeserializer implements Deserializer<Void> {
+    @Override public Void deserialize(String topic, byte[] data) {
```
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/common/serialization/VoidDeserializer.java]: nit: put `@Override` on its own line

File: clients/src/main/java/org/apache/kafka/common/serialization/VoidSerializer.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.serialization;
+
+public class VoidSerializer implements Serializer<Void> {
+    @Override public byte[] serialize(String topic, Void data) {
```
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/common/serialization/VoidSerializer.java]: nit: same as above

File: clients/src/test/java/org/apache/kafka/common/serialization/SerializationTest.java [hunk 3]
```diff
@@ -147,6 +147,13 @@ public void floatSerdeShouldPreserveNaNValues() {
         }
     }
 
+    @Test(expected = IllegalArgumentException.class)
```
Comment (ableegoldman) [inline clients/src/test/java/org/apache/kafka/common/serialization/SerializationTest.java:164]: Can we add a unit test for serialization as well, and maybe one for "normal" deserialization (where no exception is thrown)?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 4]
```diff
@@ -1878,4 +1878,34 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     assertTrue("Expected consumer2 to consume one message from offset 1, which is the committed offset of consumer1",
       records2.count() == 1 && records2.records(tp).asScala.head.offset == 1)
   }
+
+  @Test
+  def testVoidSerdeConsumption(): Unit = {
```
Comment (ableegoldman) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Do we really need a scala integration test for this? It seems like it should be covered well enough by the unit tests you wrote

**Review by ableegoldman — COMMENTED** [review]
Thanks! It mostly LGTM, just a few minor comments

File: clients/src/main/java/org/apache/kafka/common/serialization/VoidDeserializer.java (hunk 1 above)
Comment (nizhikov) [reply clients/src/main/java/org/apache/kafka/common/serialization/VoidDeserializer.java]: Fixed.

File: clients/src/test/java/org/apache/kafka/common/serialization/SerializationTest.java (hunk 3 above)
Comment (nizhikov) [reply clients/src/test/java/org/apache/kafka/common/serialization/SerializationTest.java:164]: Fixed. Two tests added - `testSerializeVoid`, `testDeserializeVoid`

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 4 above)
Comment (nizhikov) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: `testVoidSerdeConsumption` removed.

**Review by ableegoldman — APPROVED** [review]
Everything looks good, thanks!

**Comment by nizhikov** [comment]
Hello, @vvcephei Can you. please, join to the review.

Tests are green.
Sophie accepted PR.

File: clients/src/test/java/org/apache/kafka/common/serialization/SerializationTest.java [hunk 5]
```diff
@@ -147,6 +147,27 @@ public void floatSerdeShouldPreserveNaNValues() {
         }
     }
 
+    @Test
+    public void testSerializeVoid() {
+        try (Serde<Void> serde = Serdes.Void()) {
```
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/common/serialization/SerializationTest.java:152]: Why do we need a `try` for this case? The `close()` is a no-op anyway. (Similar below)

**Comment by mjsax** [comment]
Thanks for the KIP and the PR @nizhikov!

**Comment by nizhikov** [comment]
@mjsax @ableegoldman 

Thasnk you, very much for the review and merge!

**Pull request merged**